src/node_jsvmapi.cc

```diff
@@ -465,22 +465,58 @@ namespace v8impl {
 
 }  // end of namespace v8impl
 
-// We define a macro here because, with full error handling, we may want to
-// perform additional actions at the top of our public APIs, such as bail,
-// returning an error code at the very top of each public API in the case that
-// there's already an exception pending in v8impl::TryCatch::lastException
+   // We define a macro here because, with full error handling, we may want to
+   // perform additional actions at the top of our public APIs, such as bail,
+   // returning an error code at the very top of each public API in the case that
+   // there's already an exception pending in v8impl::TryCatch::lastException
 #define NAPI_PREAMBLE(e) \
```

Comment:
Yes, it should. Additionally, if there's a pending exception, it should bail early. So, with the macro I proposed [below](https://github.com/nodejs/abi-stable-node/pull/59#discussion_r97736487), this would look like
```C++
#define NAPI_PREAMBLE(e)                                            \
do {                                                                \
  RETURN_VALUE_IF_FAIL(v8impl::TryCatch::lastException().IsEmpty(), \
    napi_pending_exception);                                        \
  napi_clear_last_error_code();                                     \
  v8impl::TryCatch tryCatch(v8impl::V8IsolateFromJsEnv((e)));       \
} while(0)
```
So, if you call an API, that throws an exception, and you call another API, you lose whatever the original error code was, and the new error code becomes `napi_exception_pending`, and that's what you're gonna get from every subsequent API call.

BTW, I'm also proposing a new `napi_errorvalue` here :)